sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala

```diff
@@ -220,88 +233,151 @@ class BulkWriter(container: CosmosAsyncContainer,
+                //scalastyle:off magic.number
+                val maxItemOperationsToShowInErrorMessage = 10
+                //scalastyle:on magic.number
+                activeOperationsSnapshot
+                  .take(maxItemOperationsToShowInErrorMessage)
+                  .foreach(itemOperation => {
+                    if (sb.nonEmpty) {
+                      sb.append(", ")
+                    }
+
+                    sb.append(itemOperation.getOperationType)
+                    sb.append("->")
+                    val ctx = itemOperation.getContext[OperationContext]
+                    sb.append(s"${ctx.partitionKeyValue}/${ctx.itemId}/${ctx.eTag}(${ctx.attemptNumber})")
+                  })
+
+                if (activeOperationsSnapshot.equals(activeOperations)) {
+                  numberOfIntervalsWithIdenticalActiveOperationSnapshots += 1
+                  log.logWarning(
+                    s"FlushAndClose has been waiting ${numberOfIntervalsWithIdenticalActiveOperationSnapshots} " +
+                      s"times for identical set of operations: ${sb} Context: ${operationContext.toString}"
+                  )
+                } else {
+                  numberOfIntervalsWithIdenticalActiveOperationSnapshots = 0
+                  log.logInfo(
+                    s"FlushAndClose is waiting for active operations: ${sb} Context: ${operationContext.toString}"
+                  )
+                }
 
-        // error handling, if there is any error and the subscription is cancelled
-        // the remaining tasks will not be processed hence we never reach activeTasks = 0
-        // once we do error handling we should think how to cover the scenario.
-        lock.lock()
-        try {
-          var activeTasksSnapshot = activeTasks.get()
-          while (activeTasksSnapshot > 0 || errorCaptureFirstException.get != null) {
-            log.logDebug(s"Waiting for pending activeTasks $activeTasksSnapshot, Context: ${operationContext.toString}")
-            pendingTasksCompleted.await()
-            activeTasksSnapshot = activeTasks.get()
-            log.logDebug(s"Waiting completed for pending activeTasks $activeTasksSnapshot, Context: ${operationContext.toString}")
+                if (numberOfIntervalsWithIdenticalActiveOperationSnapshots >= 15) {
+                  errorCaptureFirstException.set(new IllegalStateException(
+                    s"Stale bulk ingestion identified - the following active operations have not been completed " +
+                      s"(first ${maxItemOperationsToShowInErrorMessage} shown) or progressed after 15 minutes: $sb"
+                  ))
+                }
+              }
+              activeTasksSnapshot = activeTasks.get()
+              val semaphoreAvailablePermitsSnapshot = semaphore.availablePermits()
+              val toBeReleased = maxPendingOperations - activeTasksSnapshot - semaphoreAvailablePermitsSnapshot
+              if (toBeReleased > 0) {
```

Comment:
why we need to release the semaphore here? it should always be 0?  or it is just double protection?

Reply:
This is within the while loop - in one of the hangs I saw that the thread was blocked on trying to acquire semaphore.

So releasing slots on the semaphore after each 1 minute wait in flushAndClose is just an attempt to prevent that situation. To be honest I haven't found evidence that we would actually leak semaphore slots - so this is more a plan-b. But I don't think it would ever harm - and with the warning log it might at least help in investigations going forward.